Enemies/Enemy.gd

```diff
@@ -143,74 +161,111 @@ var rng = RandomNumberGenerator.new()
 var cover_timer = 0
 var cover_timer_limit = 3
 
+func check_goal() -> bool:
+	return $GOAP_Planner.check_current_goal()
+
+func replan_actions():
+	state = IDLE
+
+func ready_for_action():
+	state = TAKE_ACTION
+
+func go_to_next_action():
+	action_index = action_index + 1
+	state = MOVE_TO
+	if action_index > action_plan.size() - 1:
+		action_index = 0
+		state = IDLE
+
 func _process(delta):
+	
+	#if state != IDLE and ($ReplanTimer.get_time_left() == 0 or not $GOAP_Planner.check_current_goal(world_state)):
+	if state != IDLE and not $GOAP_Planner.check_current_goal(world_state):
+		state = IDLE
+		action_index = 0
+		rng.randomize()
+		$ReplanTimer.set_wait_time(rng.randf_range(3.5, 5.0))
+		$ReplanTimer.start()
+	
 	match state:
-		FIND:
-			if can_see_player:
-				cover_timer += delta
-				var distance = translation.distance_to(player.translation)
-				if distance <= 10:
-					state = FIND_COVER
-					cover_timer = 0
-					return
-			if cover_timer > 3:
-				cover_timer = 0
-				state = FIND_COVER
-				return
-			aim_at_player(delta)
-			update_path(player.translation)
-			move_along_path(delta)
-
-#				path.clear()
-		HOLD:
-			aim_at_player(delta)
-			var distance = translation.distance_to(player.translation)
-			if distance > 10:
-				state = FIND
-		PATROL:
-			if check_vision():
-				alert_comrades()
-				state = FIND_COVER
-				clear_node_data()
-				return
-			if $PatrolTimer.get_time_left() > 0:
-				return
-			else:
-				prep_node(patrolNodes[patrolNodeIndex])
-				move_along_path(delta, true)
-				if path.size() < 1:
-					prep_node(patrolNodes[patrolNodeIndex])
-					if patrolNodeIndex == patrolNodes.size() - 1:
-						patrolNodeIndex = 0
-					else:
-						patrolNodeIndex += 1
-					$PatrolTimer.start()
-		FIND_COVER:
-			prep_node(get_shortest_node())
-			
-			state = TAKE_COVER
-		TAKE_COVER:
-			if currentNode.visible_to_player:
-				state = FIND_COVER
-				return
-			move_along_path(delta)
-			aim_at_player(delta)
-			if path.empty():
-				state = SHOOT
-				return
-		SHOOT:
-			###TO DO: ADD POPPING OUT OF COVER###
-			if can_see_player and not $Enemy_audio_player.playing():
-				$Enemy_audio_player.play_sound($Enemy_audio_player.enemy_shot)
-			aim_at_player(delta)
-			cover_timer += delta
-			if cover_timer > cover_timer_limit:
-				state = FIND
-				cover_timer = 0
-				return
-			if currentNode.visible_to_player:
-				state = FIND_COVER
-				return
+		IDLE:
+			action_plan = $GOAP_Planner.plan_actions(world_state)
+			action_index = 0
+			state = MOVE_TO
+		MOVE_TO:
+			action_plan[action_index].move_to(self, delta)
+		TAKE_ACTION:
+			action_plan[action_index].take_action(self, delta)
+	
+	return
+#	match state:
+#		FIND:
```

Comment:
we can remove this commented-out stuff from the old AI, if we need it later we can just look in git history